lib/stdlib/src/supervisor.erl

```diff
@@ -334,35 +321,34 @@ init_dynamic(_State, StartSpec) ->
 
 %%-----------------------------------------------------------------
 %% Func: start_children/2
-%% Args: Children = [child_rec()] in start order
+%% Args: Children = children() % Ids in start order
 %%       SupName = {local, atom()} | {global, atom()} | {pid(), Mod}
-%% Purpose: Start all children.  The new list contains #child's
+%% Purpose: Start all children.  The new map contains #child's
 %%          with pids.
 %% Returns: {ok, NChildren} | {error, NChildren, Reason}
-%%          NChildren = [child_rec()] in termination order (reversed
-%%                        start order)
+%%          NChildren = children() % Ids in termination order
+%%                                   (reversed start order)
 %%-----------------------------------------------------------------
-start_children(Children, SupName) -> start_children(Children, [], SupName).
-
-start_children([Child|Chs], NChildren, SupName) ->
-    case do_start_child(SupName, Child) of
-	{ok, undefined} when Child#child.restart_type =:= temporary ->
-	    start_children(Chs, NChildren, SupName);
-	{ok, Pid} ->
-	    start_children(Chs, [Child#child{pid = Pid}|NChildren], SupName);
-	{ok, Pid, _Extra} ->
-	    start_children(Chs, [Child#child{pid = Pid}|NChildren], SupName);
-	{error, Reason} ->
-	    report_error(start_error, Reason, Child, SupName),
-	    {error, lists:reverse(Chs) ++ [Child | NChildren],
-	     {failed_to_start_child,Child#child.name,Reason}}
-    end;
-start_children([], NChildren, _SupName) ->
-    {ok, NChildren}.
+start_children(Children, SupName) ->
+    Start =
+        fun(Id,Child) ->
+                case do_start_child(SupName, Child) of
+                    {ok, undefined} when ?is_temporary(Child) ->
+                        remove;
+                    {ok, Pid} ->
+                        {update,Child#child{pid = Pid}};
+                    {ok, Pid, _Extra} ->
+                        {update,Child#child{pid = Pid}};
+                    {error, Reason} ->
+                        report_error(start_error, Reason, Child, SupName),
+                        {abort,{failed_to_start_child,Id,Reason}}
+                end
+        end,
+    children_map(Start,Children).
```

Comment:
This could be `children_map(fun attempt_start_child/2, Children).`

Reply:
:+1:

Reply:
See [this comment](https://github.com/erlang/otp/pull/1602#discussion_r149643283).

Reply:
Yeah I see your function closes over SupName. Thanks for considering it.